# This Script contains:
1. Import of Libraries and Data set
2. Data Consistency Check
3. Data Cleaning
4. Export of Clean data

# 1. Import of Libraries and Data set

In [38]:
#Importing libraries
import pandas as pd
import numpy as np
import os

In [40]:
#Import Dataset
df=pd.read_csv(r'/Users/navneetjha/Desktop/Career foundry Docs/exercises/Data Immersion /Achievement 6/Dataset/gun-violence-data_01-2013_03-2018.csv',index_col=False)

In [41]:
# Checking head rows of the data set
df.head(50)

incident_id        date                 state        city_or_county  \
0        461105  2013-01-01          Pennsylvania            Mckeesport   
1        460726  2013-01-01            California             Hawthorne   
2        478855  2013-01-01                  Ohio                Lorain   
3        478925  2013-01-05              Colorado                Aurora   
4        478959  2013-01-07        North Carolina            Greensboro   
5        478948  2013-01-07              Oklahoma                 Tulsa   
6        479363  2013-01-19            New Mexico           Albuquerque   
7        479374  2013-01-21             Louisiana           New Orleans   
8        479389  2013-01-21            California             Brentwood   
9        492151  2013-01-23              Maryland             Baltimore   
10       491674  2013-01-23             Tennessee           Chattanooga   
11       479413  2013-01-25              Missouri           Saint Louis   
12       479561  2013-01-26             Louisiana             Charenton   
13       479554  2013-01-26  District of Columbia            Washington   
14       479460  2013-01-26                  Ohio           Springfield   
15       479573  2013-02-02             Tennessee               Memphis   
16       479580  2013-02-03            California         Yuba (county)   
17       479592  2013-02-07              Illinois               Chicago   
18       479603  2013-02-09             Louisiana           New Orleans   
19       480311  2013-02-11            California               Vallejo   
20       480327  2013-02-11              Delaware            Wilmington   
21       480344  2013-02-12                  Utah               Midvale   
22       480358  2013-02-19            California       Orange (county)   
23       480383  2013-02-21              Oklahoma                 Tulsa   
24       480401  2013-02-22              Michigan          Grand Rapids   
25       480407  2013-02-23            California             Lancaster   
26       480443  2013-02-24               Georgia                 Macon   
27       481186  2013-03-02             Louisiana            Shreveport   
28       481198  2013-03-03               Georgia              Moultrie   
29       481208  2013-03-03              Michigan      Saginaw (county)   
30       481213  2013-03-04            California             Los Banos   
31       481220  2013-03-05               Indiana          Indianapolis   
32       481229  2013-03-07           Mississippi               Jackson   
33       481237  2013-03-10              Missouri           Kansas City   
34       482771  2013-03-11  District of Columbia            Washington   
35       482801  2013-03-13            California             Oceanside   
36       482856  2013-03-13              New York                Mohawk   
37       482838  2013-03-14            California               Modesto   
38       482926  2013-03-16            California                  Galt   
39       482934  2013-03-17            California              Stockton   
40       482942  2013-03-17               Florida           Belle Glade   
41       483737  2013-03-21              Illinois               Chicago   
42       483765  2013-03-21              Missouri           Kansas City   
43       483788  2013-03-22              New York              Brooklyn   
44       483835  2013-03-30            California       Merced (county)   
45       483817  2013-03-31            Washington                Auburn   
46       483861  2013-04-06        South Carolina             Greenwood   
47       483872  2013-04-07              New York  New York (Manhattan)   
48       483878  2013-04-07            California            Long Beach   
49       484268  2013-04-09          Pennsylvania          Philadelphia   

                                              address  n_killed  n_injured  \
0           1506 Versailles Avenue and Coursin Street         0          4   
1                     

# 2. Data Consistency Check

In [44]:
# Checking shape of the data set
df.shape

(239677, 29)

###### There are 239677 rows and 29 columns in the raw data set

In [47]:
# Checking data info
df.info

<bound method DataFrame.info of         incident_id        date           state city_or_county  \
0            461105  2013-01-01    Pennsylvania     Mckeesport   
1            460726  2013-01-01      California      Hawthorne   
2            478855  2013-01-01            Ohio         Lorain   
3            478925  2013-01-05        Colorado         Aurora   
4            478959  2013-01-07  North Carolina     Greensboro   
...             ...         ...             ...            ...   
239672      1083142  2018-03-31       Louisiana          Rayne   
239673      1083139  2018-03-31       Louisiana   Natchitoches   
239674      1083151  2018-03-31       Louisiana         Gretna   
239675      1082514  2018-03-31           Texas        Houston   
239676      1081940  2018-03-31           Maine   Norridgewock   

                                          address  n_killed  n_injured  \
0       1506 Versailles Avenue and Coursin Street         0          4   
1                    13500 

In [49]:
# Checking data description
df.describe()

incident_id       n_killed      n_injured  congressional_district  \
count  2.396770e+05  239677.000000  239677.000000           227733.000000   
mean   5.593343e+05       0.252290       0.494007                8.001265   
std    2.931287e+05       0.521779       0.729952                8.480835   
min    9.211400e+04       0.000000       0.000000                0.000000   
25%    3.085450e+05       0.000000       0.000000                2.000000   
50%    5.435870e+05       0.000000       0.000000                5.000000   
75%    8.172280e+05       0.000000       1.000000               10.000000   
max    1.083472e+06      50.000000      53.000000               53.000000   

            latitude      longitude  n_guns_involved  state_house_district  \
count  231754.000000  231754.000000    140226.000000         200905.000000   
mean       37.546598     -89.338348         1.372442             55.447132   
std         5.130763      14.359546         4.678202             42.048117   
min        19.111400    -171.429000         1.000000              1.000000   
25%        33.903400     -94.158725         1.000000             21.000000   
50%        38.570600     -86.249600         1.000000             47.000000   
75%        41.437375     -80.048625         1.000000             84.000000   
max        71.336800      97.433100       400.000000            901.000000   

       state_senate_district  
count           207342.00000  
mean                20.47711  
std                 14.20456  
min                  1.00000  
25%                  9.00000  
50%                 19.00000  
75%                 30.00000  
max                 94.00000

In [51]:
# Checking basis stats
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 239677 entries, 0 to 239676
Data columns (total 29 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   incident_id                  239677 non-null  int64  
 1   date                         239677 non-null  object 
 2   state                        239677 non-null  object 
 3   city_or_county               239677 non-null  object 
 4   address                      223180 non-null  object 
 5   n_killed                     239677 non-null  int64  
 6   n_injured                    239677 non-null  int64  
 7   incident_url                 239677 non-null  object 
 8   source_url                   239209 non-null  object 
 9   incident_url_fields_missing  239677 non-null  bool   
 10  congressional_district       227733 non-null  float64
 11  gun_stolen                   140179 non-null  object 
 12  gun_type                     140226 non-null  object 
 13 

Multiple columns have missing values

In [54]:
# Checking data columns
df.columns

Index(['incident_id', 'date', 'state', 'city_or_county', 'address', 'n_killed',
       'n_injured', 'incident_url', 'source_url',
       'incident_url_fields_missing', 'congressional_district', 'gun_stolen',
       'gun_type', 'incident_characteristics', 'latitude',
       'location_description', 'longitude', 'n_guns_involved', 'notes',
       'participant_age', 'participant_age_group', 'participant_gender',
       'participant_name', 'participant_relationship', 'participant_status',
       'participant_type', 'sources', 'state_house_district',
       'state_senate_district'],
      dtype='object')

In [56]:
# Extracting relevant columns
df_2=df[['incident_id','date','state','city_or_county','address', 'n_killed', 'n_injured','congressional_district', 'gun_stolen','gun_type',
         'incident_characteristics', 'latitude','location_description', 'longitude', 'n_guns_involved','participant_age', 'participant_age_group', 'participant_gender','participant_relationship', 'participant_status',
         'participant_type','state_house_district', 'state_senate_district']]

In [58]:
#Printing head rows of df_2
df_2.head()

incident_id        date           state city_or_county  \
0       461105  2013-01-01    Pennsylvania     Mckeesport   
1       460726  2013-01-01      California      Hawthorne   
2       478855  2013-01-01            Ohio         Lorain   
3       478925  2013-01-05        Colorado         Aurora   
4       478959  2013-01-07  North Carolina     Greensboro   

                                     address  n_killed  n_injured  \
0  1506 Versailles Avenue and Coursin Street         0          4   
1               13500 block of Cerise Avenue         1          3   
2                      1776 East 28th Street         1          3   
3           16000 block of East Ithaca Place         4          0   
4                  307 Mourning Dove Terrace         2          2   

   congressional_district              gun_stolen                gun_type  \
0                    14.0                     NaN                     NaN   
1                    43.0                     NaN                     NaN   
2                     9.0  0::Unknown||1::Unknown  0::Unknown||1::Unknown   
3                     6.0                     NaN                     NaN   
4                     6.0  0::Unknown||1::Unknown  0::Handgun||1::Handgun   

   ... longitude  n_guns_involved                    participant_age  \
0  ...  -79.8559              NaN                              0::20   
1  ... -118.3330              NaN                              0::20   
2  ...  -82.1377              2.0  0::25||1::31||2::33||3::34||4::33   
3  ... -104.8020              NaN         0::29||1::33||2::56||3::33   
4  ...  -79.9569              2.0         0::18||1::46||2::14||3::47   

                               participant_age_group  \
0  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
1  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
2  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
3  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
4  0::Adult 18+||1::Adult 18+||2::Teen 12-17||3::...   

                            participant_gender participant_relationship  \
0         0::Male||1::Male||3::Male||4::Female                      NaN   
1                                      0::Male                      NaN   
2  0::Male||1::Male||2::Male||3::Male||4::Male                      NaN   
3         0::Female||1::Male||2::Male||3::Male                      NaN   
4       0::Female||1::Male||2::Male||3::Female                3::Family   

                                  participant_status  \
0  0::Arrested||1::Injured||2::Injured||3::Injure...   
1      0::Killed||1::Injured||2::Injured||3::Injured   
2  0::Injured, Unharmed, Arrested||1::Unharmed, A...   
3         0::Killed||1::Killed||2::Killed||3::Killed   
4       0::Injured||1::Injured||2::Killed||3::Killed   

                                    participant_type state_house_district  \
0  0::Victim||1::Victim||2::Victim||3::Victim||4:...                  NaN   
1  0::Victim||1::Victim||2::Victim||3::Victim||4:...                 62.0   
2  0::Subject-Suspect||1::Subject-Suspect||2::Vic...                 56.0   
3  0::Victim||1::Victim||2::Victim||3::Subject-Su...                 40.0   
4  0::Victim||1::Victim||2::Victim||3::Subject-Su...                 62.0   

  state_senate_district  
0                   NaN  
1                  35.0  
2                  13.0  
3                  28.0  
4                  27.0  

[5 rows x 23 columns]

# 3. Data Cleaning

In [110]:
#Checking for duplicate data 
df_2[df_2.duplicated()]

Empty DataFrame
Columns: [incident_id, date, state, city_or_county, address, n_killed, n_injured, congressional_district, gun_stolen, gun_type, incident_characteristics, latitude, location_description, longitude, n_guns_involved, participant_age, participant_age_group, participant_gender, participant_relationship, participant_status, participant_type, state_house_district, state_senate_district]
Index: []

[0 rows x 23 columns]

In [63]:
# Checking for missing values
df_missing_values=df_2.isnull().sum()

In [65]:
# Printing missing values
df_missing_values

incident_id                      0
date                             0
state                            0
city_or_county                   0
address                      16497
n_killed                         0
n_injured                        0
congressional_district       11944
gun_stolen                   99498
gun_type                     99451
incident_characteristics       326
latitude                      7923
location_description        197588
longitude                     7923
n_guns_involved              99451
participant_age              92298
participant_age_group        42119
participant_gender           36362
participant_relationship    223903
participant_status           27626
participant_type             24863
state_house_district         38772
state_senate_district        32335
dtype: int64

In [67]:
# Checking percentage of missing value
df_missing_value_percentage=df_2.isnull().mean()*100

In [69]:
#Printing missing value percentage
df_missing_value_percentage

incident_id                  0.000000
date                         0.000000
state                        0.000000
city_or_county               0.000000
address                      6.883013
n_killed                     0.000000
n_injured                    0.000000
congressional_district       4.983373
gun_stolen                  41.513370
gun_type                    41.493760
incident_characteristics     0.136016
latitude                     3.305699
location_description        82.439283
longitude                    3.305699
n_guns_involved             41.493760
participant_age             38.509327
participant_age_group       17.573234
participant_gender          15.171251
participant_relationship    93.418643
participant_status          11.526346
participant_type            10.373544
state_house_district        16.176771
state_senate_district       13.491073
dtype: float64

Except six columns all other columns contain missing values. 

In [72]:
#Dropping columns that have more than 80% missing values
df_3= df_2[['incident_id','date','state','city_or_county','address', 'n_killed', 'n_injured','congressional_district', 'gun_stolen','gun_type',
         'incident_characteristics', 'latitude','longitude', 'n_guns_involved','participant_age', 'participant_age_group', 'participant_gender', 'participant_status',
         'participant_type','state_house_district', 'state_senate_district']]

In [74]:
# Dropping missing values
df_cleaned=df_3.dropna()

In [76]:
#Printing cleaned dataset
df_cleaned.head()

incident_id        date           state city_or_county  \
2        478855  2013-01-01            Ohio         Lorain   
4        478959  2013-01-07  North Carolina     Greensboro   
6        479363  2013-01-19      New Mexico    Albuquerque   
10       491674  2013-01-23       Tennessee    Chattanooga   
11       479413  2013-01-25        Missouri    Saint Louis   

                                address  n_killed  n_injured  \
2                 1776 East 28th Street         1          3   
4             307 Mourning Dove Terrace         2          2   
6                        2806 Long Lane         5          0   
10                       1501 Dodds Ave         1          3   
11  W Florissant Ave and Riverview Blvd         1          3   

    congressional_district              gun_stolen  \
2                      9.0  0::Unknown||1::Unknown   
4                      6.0  0::Unknown||1::Unknown   
6                      1.0  0::Unknown||1::Unknown   
10                     3.0              0::Unknown   
11                     1.0              0::Unknown   

                        gun_type  ... latitude  longitude  n_guns_involved  \
2         0::Unknown||1::Unknown  ...  41.4455   -82.1377              2.0   
4         0::Handgun||1::Handgun  ...  36.1140   -79.9569              2.0   
6   0::22 LR||1::223 Rem [AR-15]  ...  34.9791  -106.7160              2.0   
10                    0::Unknown  ...  35.0221   -85.2697              1.0   
11                    0::Unknown  ...  38.7067   -90.2494              1.0   

                          participant_age  \
2       0::25||1::31||2::33||3::34||4::33   
4              0::18||1::46||2::14||3::47   
6   0::51||1::40||2::9||3::5||4::2||5::15   
10                                  0::19   
11                                  0::28   

                                participant_age_group  \
2   0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
4   0::Adult 18+||1::Adult 18+||2::Teen 12-17||3::...   
6   0::Adult 18+||1::Adult 18+||2::Child 0-11||3::...   
10                                       0::Adult 18+   
11                                       0::Adult 18+   

                                   participant_gender  \
2         0::Male||1::Male||2::Male||3::Male||4::Male   
4              0::Female||1::Male||2::Male||3::Female   
6   0::Male||1::Female||2::Male||3::Female||4::Fem...   
10                 0::Male||1::Male||2::Male||3::Male   
11                                            0::Male   

                                   participant_status  \
2   0::Injured, Unharmed, Arrested||1::Unharmed, A...   
4        0::Injured||1::Injured||2::Killed||3::Killed   
6   0::Killed||1::Killed||2::Killed||3::Killed||4:...   
10      0::Killed||1::Injured||2::Injured||3::Injured   
11      0::Killed||1::Injured||2::Injured||3::Injured   

                                     participant_type state_house_district  \
2   0::Subject-Suspect||1::Subject-Suspect||2::Vic...                 56.0   
4   0::Victim||1::Victim||2::Victim||3::Subject-Su...                 62.0   
6   0::Victim||1::Victim||2::Victim||3::Victim||4:...                 10.0   
10  0::Victim||1::Victim||2::Victim||3::Victim||4:...                 28.0   
11  0::Victim||1::Victim||2::Victim||3::Victim||4:...                 76.0   

    state_senate_district  
2                    13.0  
4                    27.0  
6                    14.0  
10                   10.0  
11                    4.0  

[5 rows x 21 columns]

In [86]:
# Checking shape of cleaned data set
df_cleaned.shape

(67116, 21)

In [90]:
df_cleaned.value_counts().sum()

67116

In [100]:
#Checking for missing values in cleaned set
df_cleaned.isnull().sum()

incident_id                 0
date                        0
state                       0
city_or_county              0
address                     0
n_killed                    0
n_injured                   0
congressional_district      0
gun_stolen                  0
gun_type                    0
incident_characteristics    0
latitude                    0
longitude                   0
n_guns_involved             0
participant_age             0
participant_age_group       0
participant_gender          0
participant_status          0
participant_type            0
state_house_district        0
state_senate_district       0
dtype: int64

In [102]:
# Checking basic info about cleaned data
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 67116 entries, 2 to 239676
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   incident_id               67116 non-null  int64  
 1   date                      67116 non-null  object 
 2   state                     67116 non-null  object 
 3   city_or_county            67116 non-null  object 
 4   address                   67116 non-null  object 
 5   n_killed                  67116 non-null  int64  
 6   n_injured                 67116 non-null  int64  
 7   congressional_district    67116 non-null  float64
 8   gun_stolen                67116 non-null  object 
 9   gun_type                  67116 non-null  object 
 10  incident_characteristics  67116 non-null  object 
 11  latitude                  67116 non-null  float64
 12  longitude                 67116 non-null  float64
 13  n_guns_involved           67116 non-null  float64
 14  participan

In [92]:
# Checking basis description of cleaned data
df_cleaned.describe()

incident_id      n_killed     n_injured  congressional_district  \
count  6.711600e+04  67116.000000  67116.000000            67116.000000   
mean   7.220504e+05      0.367081      0.503084                8.030052   
std    2.427656e+05      0.638799      0.784947                8.217204   
min    9.213500e+04      0.000000      0.000000                0.000000   
25%    5.463350e+05      0.000000      0.000000                2.000000   
50%    7.655990e+05      0.000000      0.000000                5.000000   
75%    9.161165e+05      1.000000      1.000000               11.000000   
max    1.083435e+06     50.000000     53.000000               53.000000   

           latitude     longitude  n_guns_involved  state_house_district  \
count  67116.000000  67116.000000     67116.000000          67116.000000   
mean      37.765321    -89.700344         1.292196             53.334019   
std        4.847671     14.119407         3.328392             43.727232   
min       19.112700   -159.412000         1.000000              1.000000   
25%       34.165375    -92.337925         1.000000             19.000000   
50%       39.045600    -87.054000         1.000000             43.000000   
75%       41.697425    -80.785975         1.000000             80.000000   
max       61.229600    -67.271100       400.000000            814.000000   

       state_senate_district  
count           67116.000000  
mean               20.113982  
std                14.373291  
min                 1.000000  
25%                 8.000000  
50%                18.000000  
75%                30.000000  
max                94.000000

It looks like the column n_guns_involved contains an outlier.

In [108]:
#Checking for duplicate data in cleaned data set
df_cleaned[df_cleaned.duplicated()]

Empty DataFrame
Columns: [incident_id, date, state, city_or_county, address, n_killed, n_injured, congressional_district, gun_stolen, gun_type, incident_characteristics, latitude, longitude, n_guns_involved, participant_age, participant_age_group, participant_gender, participant_status, participant_type, state_house_district, state_senate_district]
Index: []

[0 rows x 21 columns]

In [98]:
# Exporting cleaned data set 
df_cleaned.to_csv('gun_violence_cleaned',index=False)